SYCL/ESIMD/regression/complex-lib-lin.cpp

```diff
@@ -0,0 +1,36 @@
+// This test is intended to check a certain approach of compiling libraries and
+// application, when both regular SYCL and ESIMD are used.
+//
+// We used to have a bug, when under some circumstances compiler created empty
+// device images, but at the same time it stated that they contain some kernels.
+// More details can be found in intel/llvm#4927.
+//
+// REQUIRES: linux && (cpu || gpu)
+// UNSUPPORTED: cuda || hip
+//
+// RUN: %clangxx -fsycl -fPIC -O3 %S/Inputs/complex-lib-sycl.cpp -c -o %t-lib-sycl.o
+// RUN: %clangxx -fsycl -fPIC -O3 %S/Inputs/complex-lib-esimd.cpp -c -o %t-lib-esimd.o
+// RUN: %clangxx -fsycl -fPIC -O3 %S/Inputs/complex-lib-test.cpp -c -o %t-test.o
+//
+// RUN: ar crsv %t-lib-sycl.a %t-lib-sycl.o
+// RUN: ar crsv %t-lib-esimd.a %t-lib-esimd.o
+//
+// One shared library is built using static libraries
+//
+// RUN: %clangxx -fsycl -shared %t-lib-sycl.a %t-lib-esimd.a \
+// RUN:  -fsycl-device-code-split=per_kernel -Wl,--whole-archive \
+// RUN:  %t-lib-sycl.a %t-lib-esimd.a -Wl,--no-whole-archive -Wl,-soname,%S -o %t-lib-a.so
+//
+// And another one is constructed directly from object files
+//
+// RUN: %clangxx -fsycl -shared %t-lib-sycl.o %t-lib-esimd.o \
+// RUN:  -fsycl-device-code-split=per_kernel -Wl,-soname,%S -o %t-lib-o.so
+//
+// RUN: %clangxx -fsycl %t-test.o %t-lib-a.so -o %t-a.run
+// RUN: %clangxx -fsycl %t-test.o %t-lib-o.so -o %t-o.run
+//
+// FIXME: is there better way to handle libraries loading than LD_PRELOAD?
```

Comment:
Please comment why special handling is needed and simple LD_LIBRARY_PATH is not enough?

Reply:
The thing is that those `.so` libraries are generated using some temporary names/paths provided by LIT infrastructure: I simply wasn't able to find in the documentation how to reliably get path to a _directory_ with those libraries to use `LD_LIBRARY_PATH`

Reply:
Can the output directory be forced at compile-time and reused at runtime? E.g.
```
%clangxx -fsycl -shared ... -o %T/lib-a.so
...
%GPU_RUN_PLACEHOLDER LD_LIBRARY_PATH=%T:$LD_LIBRARY_PATH %t-o.run
```

Reply:
> Can the output directory be forced at compile-time and reused at runtime? 

The main issue is how to obtain a path to a (preferably) unique directory, which will be automatically removed after tests execution?

There is a `%T` built-in [substitution](https://llvm.org/docs/CommandGuide/lit.html#substitutions), but it is marked as "do not use":

> %T | parent directory of %t (not unique, deprecated, do not use)

Reply:
Added info about `%T` into the test source in c18962fa70c8900fd6aa9022cd3f75327ebe137d